hive-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/HiveBatchSink.java

```diff
@@ -127,8 +111,54 @@ public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, HCa
   public static class SinkOutputFormatProvider implements OutputFormatProvider {
     private final Map<String, String> conf;
 
-    public SinkOutputFormatProvider() {
-      this.conf = new HashMap<>();
+    public SinkOutputFormatProvider(BatchSinkContext context, Job job, HiveSinkConfig config) throws IOException {
+      Configuration originalConf = job.getConfiguration();
+      Configuration modifiedConf = new Configuration(originalConf);
+      modifiedConf.set("hive.metastore.uris", config.metaStoreURI);
+      HCatOutputFormat.setOutput(modifiedConf, job.getCredentials(), OutputJobInfo.create(config.dbName,
+                                                                                          config.tableName,
+                                                                                          getPartitions(config)));
+
+      HCatSchema hiveSchema = HCatOutputFormat.getTableSchema(modifiedConf);
+      if (config.schema != null) {
+        // if the user did provide a sink schema to use then use that one
+        hiveSchema = HiveSchemaConverter.toHiveSchema(Schema.parseJson(config.schema), hiveSchema);
+      }
+      HCatOutputFormat.setSchema(modifiedConf, hiveSchema);
+      conf = getConfigurationDiff(originalConf, modifiedConf);
+      HiveSchemaStore.storeHiveSchema(context, config.dbName, config.tableName, hiveSchema);
```

Comment:
this should happen in the beforeSubmit method rather than in this constructor. Usually don't want to do stuff that hits external systems in a constructor.